ipyparallel/cluster/shellcmd.py

```diff
@@ -0,0 +1,860 @@
+        self.pathsep = (
+            "/"  # equivalent to os.pathsep (will be changed during initialization)
+        )
+
+        self.send_receiver_class = (
+            send_receiver_class  # should be activated when developing...
+        )
+        self.debugging = False  # for outputs to file for easier debugging
+        self.log = None
+        if log:
+            if isinstance(log, str):
+                self.log = SimpleLog(log)
+            else:
+                self.log = log
+
+        if initialize:
+            self.initialize()
+
+    @staticmethod
+    def _check_output(cmd, **kwargs):
+        output = check_output(cmd, **kwargs)
+        if isinstance(output, str):
+            return output
+        else:
+            return output.decode('utf8', 'replace')
+
+    @staticmethod
+    def _runs_successful(cmd):
+        try:
+            check_output(cmd)
+        except CalledProcessError as e:
+            return False
+        return True
+
+    @staticmethod
+    def _as_list(cmd):
+        if isinstance(cmd, str):
+            return [cmd]
+        elif isinstance(cmd, list):
+            return cmd
+        else:
+            raise Exception("Unknown command type")
+
+    @staticmethod
+    def _format_for_python(param):
+        if isinstance(param, str):
+            return f"'{param.translate(ShellCommandSend._python_chars_map)}'"
+        else:
+            return str(param)
+
+    def _cmd_send_via_shell(self, paramlist):
+        # unused/deprecated function which send a command as shell command rather than as python code (see _cmd_send)
+        # since it requires proper quoting for different shells, it's much more sensitive and complex than sending
+        # poor python code
+        def _powershell_quote(param):
+            if '"' in param or "'" in param or " " in param:
+                # we need to replace single and double quotes be two double quotes, but if we are inside a string,
+                # we need to prepend a backslash to the double quote. Otherwise it will get removed
+                quoted = ""
+                in_string = False
+                for idx, c in enumerate(param):
+                    prev_c = None if idx == 0 else param[idx - 1]
+                    next_c = None if idx == len(param) - 1 else param[idx + 1]
+                    if c == '"' and prev_c != "\\":
+                        in_string = not in_string
+                        quoted += '"' * 2
+                        continue
+                    if c == "'":
+                        if in_string:
+                            quoted += '\\"' * 2
+                        else:
+                            quoted += '"' * 2
+                        continue
+                    quoted += c
+                return "'" + quoted + "'"
+            else:
+                return param
+
+        def _cmd_quote(param):
```

Comment:
why not use [shlex](https://docs.python.org/3/library/shlex.html) utilities for splitting and joining commands? It's more likely to be robust to nested quotes, etc.